Tech bse

sangallidavide opened the pull request:
This contains part of a old pull request which was never merged in the private repo
It contains a development used to compute circular dichroism on molecules beyond IP
https://github.com/yambo-code/yambo-devel/pull/725

Specifically it contains the first two parts, and not the third point, which will be merged later to reduce the size of the merge

* Changes to reduce the memory usage of the BSE solver. The kernel subroutine `K.F` now renamed `K_kernel.F` computes the `"x"` and the `"c"` inside a loop. Similarly also the `TDDFT` part is done independently. 
```
 do iHxc=1,3
   !
   if ((.not.l_bs_exch).and.iHxc==1) cycle
   !
   if ((.not.BS_K_is_ALDA).and.iHxc==2) cycle
   !
   if ((.not.l_bs_corr).and.iHxc==3) cycle
   !
```
The main advantage is that first only the variables required for `"x"` are allocated and later only the variables required for `"c"`. Another advantage is that, by doing so, the cutoff on the correlation and the exchange part do not need to be related


* Changes to the structure of the diago residuals. 
There are now two key subroutines for the residuals:
```
K_diago_right_residuals(mode,i_BS_mat,BS_E,BS_R_right,BS_V_right)
K_diago_left_residuals(mode,i_BS_mat,BS_E,BS_R_left,BS_V_left,BS_overlap)
```
which can be used for any general kind of dipoles, and not anymore subroutines which depend on the physical quantity to be computed

Other changes:
* a new subroutine containing BSE_utilities has been created
* running bse at the IP level it is possible to get in output a file with the list of transitions for each pole of the IP response function
* K_Transitions_setup and K_Transitions_dimensions split into 2 independent subroutines
* wave_functions when evaluating exchange in BSE can be loaded in a loop reducing memory

sangallidavide committed:
Merge remote-tracking branch 'yambo-gpl/tech-master'

sangallidavide committed:
Version 5.2.0,  Revision 23181,  Hash e9610626a2

MODIFIED *  configure include/version/version.m4 bse/K_driver_init.F tddft/TDDFT_ALDA_eh_space_G_collisions_R.F

Bugs:
- [yambo] Fixed compilation with CUDAF

Changes:
- [yambo] Added error message for TDDFT eh space with G-integrals with CUDAF

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Version 5.2.0,  Revision 23182,  Hash 740f9a89c2

MODIFIED *  configure include/version/version.m4

RENAMED *  bse/K.F -> bse/K_kernel.F

Changes:
- K.F renamed K_kernel.F

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Version 5.2.0,  Revision 23183,  Hash cfbf2392ee

MODIFIED *  configure include/version/version.m4 bse/.objects bse/DOUBLE_project.dep bse/K_Haydock.F bse/K_Haydock_response.F bse/K_IP.F bse/K_IP_sort.F bse/K_Transitions_setup.F bse/K_components_folded_in_serial_arrays.F bse/K_correlation_collisions_dir.F bse/K_correlation_collisions_std.F bse/K_correlation_kernel_dir.F bse/K_diago_driver.F bse/K_diago_response_functions.F bse/K_dipoles.F bse/K_driver.F bse/K_driver_init.F bse/K_exchange_collisions.F bse/K_kernel.F bse/K_observables.F bse/K_solvers.F

NEW *  bse/BSE_utilities.F bse/K_diago_left_residuals.F bse/K_diago_overlap_matrix.F bse/K_diago_right_residuals.F

Changes:
- Imported BSE subroutines from phys-dichroism

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Version 5.2.0,  Revision 23184,  Hash 586cb1a06a

MODIFIED *  configure include/version/version.m4 allocations/X_ALLOC_elemental.F common/eval_G_minus_G.F interface/INIT.F interface/INIT_activate.F interface/INIT_load.F io/io_BSS_diago.F modules/SET_defaults.F modules/mod_BS.F modules/mod_BS_solvers.F modules/mod_R_lattice.F modules/mod_X.F modules/mod_interfaces.F modules/mod_wrapper.F modules/mod_wrapper_omp.F output/K_OUTPUT.F output/X_OUTPUT_driver.F tddft/TDDFT_ALDA_eh_space_G_collisions_L.F tddft/TDDFT_ALDA_eh_space_G_collisions_R.F wf_and_fft/WF_alloc.F wf_and_fft/WF_free.F wf_and_fft/WF_load.F

Changes:
- Imported changes to BSE structure from branch phys-dichroism

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Version 5.2.0,  Revision 23185,  Hash 0f3087daab

MODIFIED *  configure include/version/version.m4 bse/K_kernel.F interface/INIT_load.F wf_and_fft/WF_alloc.F

Changes:
- [yambo] compilation fixed

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Merge remote-tracking branch 'yambo-gpl/tech-master' into tech-bse

sangallidavide committed:
Version 5.2.0,  Revision 23233,  Hash 8979e96f27

MODIFIED *  configure include/version/version.m4 bse/DOUBLE_project.dep bse/K_IP_sort.F bse/K_observables.F collisions/DOUBLE_project.dep collisions/NL_project.dep modules/mod_BS.F output/K_OUTPUT.F output/X_OUTPUT_driver.F

DELETED *  driver/PHEL_project.dep

Changes:
- printing indexes of IP excitations must be activated in input
- fixed proj.dep files

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

attacc commented:
Just few comments:

1) Did you prepare a tutorial for  circular dichroism on molecules? and the corresponding tests?

2) It is good that one can set different number of plane waves for exchange and correlation, but one of the problem
of Yambo is that there are too much variables in the inputs (what many users say). Is it possible to do something like:

NGexchange = 1000 RL
NGcorrel        = -1 RL   ! If negative Yambo sets it equal to NGexange

put a negative value for one of the two in such a way that Yambo automatically setd it equal to the other one,
and put a higher verbosity on NGcorrel.

sangallidavide commented:
> Just few comments:
> 
>     1. Did you prepare a tutorial for  circular dichroism on molecules? and the corresponding tests?

Working on it. For now there is just an empty page: https://www.yambo-code.eu/wiki/index.php/Dichroism_in_molecules 
I'll let you know as soon as it is ready. :-)

> 
>     2. It is good that one can set different number of plane waves for exchange and correlation, but one of the problem
>        of Yambo is that there are too much variables in the inputs (what many users say). Is it possible to do something like:
> 
> 
> NGexchange = 1000 RL
> NGcorrel = -1 RL  ! If negative Yambo sets it equal to NGexange
> 
> put a negative value for one of the two in such a way that Yambo automatically setd it equal to the other one, and put a higher verbosity on NGcorrel.

I did not change the input. The two variables remain
```
BSENGexx= 4000            RL    # [BSK] Exchange components
BSENGBlk= 200            RL    # [BSK] Exchange components
```
There is already the option `BSENGBlk= -1  RL` which sets it equal to the cutoff on the screening (I think you did it, right?)

However, the present yambo will change automatically the value of `BSENGexx= 4000 RL` it the fft grid corresponding to this exchange cutoff is not compatible with the fft grid corresponding to the correlation cutoff `BSENGBlk= 200 RL`. Moreover, there is an extra variable to void this "automatically lowering", e.g. `ALLGexx`. If you use it, the lowering is not done, but the computation of the correlation part will become much slower.

With the new implementation `ALLGexx` is not needed anymore, and the cutoff is never "automatically lowered", since the eh-exchange and eh-correlation part are done independently (in a loop over the different parts of the kernel)

attacc commented:
Ok, thanks for the answers

andreamarini commented:
Hi Davide. 

I will check the source ASAP.

In the meantime: 

- why tech-? It looks to me a mixture or refact- and phys-
- I do believe that in order to enter in the GPL a piece of code describing new physics should be supported with a tutorial (you already answered to this) and a publication.

Actually regarding the publication I think that one paper is not enough. We should decide among us how and when a new physical feature is _mature_ enough to be released to the general public.

attacc reviewed (approved):
Ok for me, if tests are fine.
I notice that there are also other changes in the psuedo-potential part, coulomb cutoff, slepc,  new subroutine for scalar products, and excitons sorting.
May you briefly comment on these changes?

sangallidavide commented:
We used it for calculation on molecules,

>     * why tech-? It looks to me a mixture or refact- and phys-

There is no new physics.
Just the points mentioned above. In the meeting we agreed in having tech-master, maintentance-master and phys-master.
There is no refact-master

> 
>     * I do believe that in order to enter in the GPL a piece of code describing new physics should be supported with a tutorial (you already answered to this) and a publication.
> 
> Actually regarding the publication I think that one paper is not enough. We should decide among us how and when a new physical feature is _mature_ enough to be released to the general public.


See above, no new physcs.
It was used to compute dichroism in molecules (which can be done also with the gpl master).  However the simultaions where too demanding (expecially for memory). For this reasons the changes introduced. Related publication where this was used:
https://iopscience.iop.org/article/10.1088/1757-899X/1265/1/012005/meta
https://papers.sif.it/?pid=ncc12480
I'll also prepare a tutorial

> 
> Side question: how can I check the code if the branch is hosted on a separate repo? Do I have to clone it? Would'nt it be better to have the branch in the yambo repo so that, among other things, the robots can access it?

Yeah, you can clone it

sangallidavide commented:
> Ok for me, if tests are fine. I notice that there are also other changes in the psuedo-potential part, coulomb cutoff, 

it was merged with bug-fixes to import recent fixes


> slepc,

This are maybe the chanegs related to this point (?)
* Changes to the structure of the diago residuals. 

> new subroutine for scalar products
Some "V_dot_V"  interfaces where missing, like for r1 and r2. Do you mean this ?

> and excitons sorting. 
I guess you mean this point (?)
* running bse at the IP level it is possible to get in output a file with the list of transitions for each pole of the IP response function

attacc commented:
Thanks

sangallidavide committed:
Merge remote-tracking branch 'yambo-gpl/tech-master' into tech-bse

sangallidavide committed:
Version 5.2.0,  Revision 23237,  Hash 3c6947ba0f

MODIFIED *  configure include/version/version.m4 modules/mod_BS_solvers.F

Bugs:
- Fixed duplicated line introduced by merge

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Version 5.2.0,  Revision 23238,  Hash d491dbdf56

MODIFIED *  configure include/version/version.m4 bse/K_correlation_collisions_dir.F

Changes:
- [yambo] SET_table renamed EVAL_table to minimize changes w.r.t. the master

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

attacc commented:
Ok, since there is not new physics here, but a good optimization of memory in BSE, that one of the main iusse of Yambo, I think this branch should be merged in the tech-master

attacc commented:
But I see that there are still some test that fail
https://media.yambo-code.eu/robots/tech-bse/unimi-XPS-8930.php

sangallidavide commented:
> But I see that there are still some test that fail https://media.yambo-code.eu/robots/tech-bse/unimi-XPS-8930.php

Yes, with gfortran the tests nloptics due to the previous merge ...
https://media.yambo-code.eu/robots/tech-bse/unimi-XPS-8930_tech-bse_3_error.php

With nvfortran, CUDA and GPUs the status is the same as the tech-master.
https://media.yambo-code.eu/robots/tech-bse/unimi-XPS-8930_tech-bse_1_error.php
Here there was just an extra fail on 03_GoWo_PPA_corrections which caused subsequent fails.
There were missing dependencies via input flags, now added b1ac8c94ed21f97121c17cf27badea62135dc267

attacc commented:
I found a bug in the previous merge... fixing in progress...

sangallidavide commented:
Ok. Meanwhile tutorial for CD available: https://www.yambo-code.eu/wiki/index.php/Dichroism_in_molecules

attacc commented:
Cool!

On Fri, 21 Jun 2024 at 11:09, Davide Sangalli ***@***.***>
wrote:

> Ok. Meanwhile tutorial for CD available:
> https://www.yambo-code.eu/wiki/index.php/Dichroism_in_molecules
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yambo-code/yambo/pull/98#issuecomment-2182343126>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AED6PCIFDDZDDPSGHCGK3GLZIPUUDAVCNFSM6AAAAABJGZJU3KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBSGM2DGMJSGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
==========================================================
Mobile[FR]:    +33662922897
Mobile[IT]:     +393384315437
Fax:              +33491418916
Address: CINaM
            Campus de Luminy
            163 Avenue de Luminy, Case 913
            13288 Marseille Cedex 9

Skype: claudioattaccalite
Editor Fellow for: www.scipost.org
web site: www.attaccalite.com
===========================================================

sangallidavide committed:
Merge remote-tracking branch 'yambo-gpl/tech-master'

sangallidavide committed:
Version 5.2.0,  Revision 23255,  Hash 94511a32c9

MODIFIED *  include/version/version.m4 modules/mod_BS_solvers.F

Bugs:
- [yambo] Removed double definition of variable BSS_slepc_mpd in mod_BS_solvers, due to wrong merge

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide committed:
Merge branch 'master' into tech-bse

sangallidavide committed:
Merge remote-tracking branch 'attacc/refact_fast-collisions' into tech-bse

sangallidavide committed:
Merge remote-tracking branch 'yambo-gpl/tech-master' into tech-bse

sangallidavide committed:
Version 5.2.0,  Revision 23279,  Hash 6ac3a88014

MODIFIED *  include/version/version.m4 dipoles/DIPOLE_overlaps.F modules/mod_interfaces.F wf_and_fft/WF_shift_kpoint.F wf_and_fft/WF_symm_kpoint.F

Bugs:
- [yambo] Fixed covariant dipoles with CUDA (issue #106)

Patch sent by:  Davide Sangalli <davide.sangalli@gmail.com>

sangallidavide commented:
The branch is now green with both gfortran and nvfortran:
https://media.yambo-code.eu/robots/GPL/tech-bse/unimi-XPS-8930.php

Tutorial ready:
https://www.yambo-code.eu/wiki/index.php/Dichroism_in_molecules
(It is possible to do it also with the master, this branch just improves the memory for large runs)

Connected equations in the tutorials and also here:
https://iopscience.iop.org/article/10.1088/1757-899X/1265/1/012005/meta
https://papers.sif.it/?pid=ncc12480

The branch is both in my fork and in the main repo
https://github.com/sangallidavide/yambo/tree/tech-bse
https://github.com/yambo-code/yambo/tree/tech-bse

There has been about 1 month of time to review the code.
Unless there are other reasons for blocking it, I'd like to proceed with the merge.

attacc commented:
Fine for me, I will run also Intel tests

attacc commented:
Intel tests are fine on my PC (8 cores)
https://media.yambo-code.eu/robots/GPL/tech-bse/attacc-lavoro.php

attacc committed:
Merge branch 'tech-master' into tech-bse

attacc commented:
In the branch tech-bse I changed also the subroutine OSCLL_eval.F in order to use extreme_only
It works, maybe we can merge directly the branch instead of the fork.  I running tests again

sangallidavide commented:
I checked, but I cannot change the branch in the pull request.
For now I merged this, which also contains the discussion.
We can open a new PR for further changes.

attacc commented:
Fine for me!
